### PR TITLE
sql: implement ConstructScan in the new factory

### LINKED ISSUE #47474: execbuilder: plan TableReaderSpec directly
This is the first step towards #47473. At a high level, we should create a new implementation of an `exec.Factory` and implement `ConstructScan` to return a `TableReaderSpec` directly. This behavior should be off by default. We might want to break down this issue further or expand once we've done a bit of investigation.

### PR DESCRIPTION
**Pull request by yuzefovich**
**sql: extract out planning of table readers into a separate method**

This commit extract most of the logic of `createTableReaders` into
a separate method that performs the actual physical planning of table
readers given already created `TableReaderSpec`s. This will allow for
reuse in the follow-up work. This commit additionally slightly refactors
a few other things for reuse.

Release note: None

**sql: implement ConstructPlan by new factory**

This commit extracts the logic of `execFactory.ConstructPlan` into the
function to be used by both factories. It also extracts out another
small helper.

Release note: None

**sql: implement ConstructScan in distsql spec factory**

This commit adds an implementation of
`distSQLSpecExecFactory.ConstructScan` which combines the logic that
performs `scanNode` creation of `execFactory.ConstructScan` and physical
planning of table readers of `DistSQLPlanner.createTableReaders`.
I tried to refactor the code so that there is not much duplication going
on.

Notably, simple projections, renders, and filters are not yet
implemented.

Closes: #47474.

Release note: None

**sql: enforce the assumption that `wantedColumns` is not nil for scans**

After the examining the code closer, I see that now in all code paths we
populate `scanColumnsConfig.wantedColumns` to ask for the specific
columns to be scanned. Previously, this map could be left as `nil` which
would mean that all columns of the table should be scanned. This was
used only by the heuristic planner which has been removed, so we can now
update the assumption of the scan columns config. This allows us to
clean up the comments a bit.

This commit also moves `makeColumnsConfig` function into the shared util
file because it is used by both factories.

Release note: None

### CODE REVIEW

**Comment by yuzefovich** [comment]
cc @rytaft

**Review by rytaft — APPROVED** [review]
> Reviewed 1 of 1 files at r1, 4 of 4 files at r2, 6 of 6 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49682)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @yuzefovich)

___
*[pkg/sql/distsql_physical_planner.go, line 911 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49682#-M8vkrjn8kW1NjzJ2zr_:-M8vkrjo3ZWTS1XUyCj0:b-vd4srm) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc5e82a0abff1d13aca14648df0cc9fac7207307/pkg/sql/distsql_physical_planner.go#L911)):*
> ```Go
> }
> 
> func getColsForScanToTableOrdinalMap(
> ```

would be good to have a small comment at least
___
*[pkg/sql/distsql_physical_planner.go, line 1037 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49682#-M8vkBLM7cKyMCCWHmTD:-M8vkBLM7cKyMCCWHmTE:b-oy9neb) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc5e82a0abff1d13aca14648df0cc9fac7207307/pkg/sql/distsql_physical_planner.go#L1037)):*
> ```Go
> 	// scanNodes can have columns set up in a few different ways, depending on the
> 	// colCfg. The heuristic planner always creates scanNodes with all public
> 	// columns (even if some of them aren't even in the index we are scanning).
> ```

We don't have the heuristic planner anymore so you can take this sentence out :)

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49682)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @rytaft)

___
*[pkg/sql/distsql_physical_planner.go, line 911 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49682#-M8vkrjn8kW1NjzJ2zr_:-M8vuGIO-4-ZE0Q1-98W:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc5e82a0abff1d13aca14648df0cc9fac7207307/pkg/sql/distsql_physical_planner.go#L911)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

would be good to have a small comment at least
</blockquote></details>

Done.
___
*[pkg/sql/distsql_physical_planner.go, line 1037 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49682#-M8vkBLM7cKyMCCWHmTD:-M8w--s09Ss1AMOg5f51:b46140) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc5e82a0abff1d13aca14648df0cc9fac7207307/pkg/sql/distsql_physical_planner.go#L1037)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

We don't have the heuristic planner anymore so you can take this sentence out :)
</blockquote></details>

Good point. I examined the code a little bit more and added another commit to this PR to clean this up.

**Review by rytaft — APPROVED** [review]
:lgtm: 

> Reviewed 4 of 6 files at r6, 6 of 6 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49682#-:-M8w9RU78y98TYS6VX73:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto and @yuzefovich)

___
*[pkg/sql/exec_factory_util.go, line 90 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/49682#-M8w8-0I5Qkhb_b3nUnd:-M8w8-0I5Qkhb_b3nUne:btkudqo) ([raw file](https://github.com/cockroachdb/cockroach/blob/59ab5760a20afff6ad67675b751c066d814ae0ce/pkg/sql/exec_factory_util.go#L90)):*
> ```Go
> 	// Set visibility=execinfra.ScanVisibilityPublicAndNotPublic, since all
> 	// columns in the "cols" set should be projected, regardless of whether
> 	// they're public or non- public. The caller decides which columns to
> ```

[nit] extra space in "non- public"

**Review by yuzefovich — COMMENTED** [review]
TFTR! I see that Alfonso was also typing comments, so I'll wait for his feedback as well.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49682#-:-M8wAVQcCmgO1Ny31yFE:basldh)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto and @rytaft)

___
*[pkg/sql/exec_factory_util.go, line 90 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/49682#-M8w8-0I5Qkhb_b3nUnd:-M8wAUb71L2pdkiYwOpr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/59ab5760a20afff6ad67675b751c066d814ae0ce/pkg/sql/exec_factory_util.go#L90)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] extra space in "non- public"
</blockquote></details>

Done.

**Review by asubiotto — CHANGES_REQUESTED** [review]
Nice work! However, seeing this makes me think of how we should test it. I think something we should definitely do is differential testing of `execFactory` + `distsql physical planner` vs `distsqlSpecFactory`.

> Reviewed 1 of 1 files at r1, 9 of 9 files at r4, 4 of 4 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49682#-:-M8yhXY19ewA-eiOMjmx:b2ocjkw)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @rytaft and @yuzefovich)

___
*[pkg/sql/distsql_physical_planner.go, line 911 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49682#-M8uOxo43lvgQ2PaVCwr:-M8uOxo43lvgQ2PaVCws:b-efenpq) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc5e82a0abff1d13aca14648df0cc9fac7207307/pkg/sql/distsql_physical_planner.go#L911)):*
> ```Go
> }
> 
> func getColsForScanToTableOrdinalMap(
> ```

nit: can this function name be clearer? maybe just `getColOrdinals`?
___
*[pkg/sql/distsql_physical_planner.go, line 1063 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/49682#-M8yeCct1fsKD9nY84L9:-M8yeCcuCZyFuxKEnpFT:b-8op4pj) ([raw file](https://github.com/cockroachdb/cockroach/blob/429f61ae29994738abe4d8fd2f09c5c2f45b30ee/pkg/sql/distsql_physical_planner.go#L1063)):*
> ```Go
> 
> func (dsp *DistSQLPlanner) planTableReaders(
> 	planCtx *PlanningCtx,
> ```

These are a lot of arguments. What do you think of creating some argument struct and maybe documenting the arguments better?
___
*[pkg/sql/distsql_spec_exec_factory.go, line 60 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/49682#-M8yhDhY9Iramwdn9Mct:-M8yhDhZEbrTMU9TUSQK:b-gwpf7m) ([raw file](https://github.com/cockroachdb/cockroach/blob/429f61ae29994738abe4d8fd2f09c5c2f45b30ee/pkg/sql/distsql_spec_exec_factory.go#L60)):*
> ```Go
> 	rowCount float64,
> 	locking *tree.LockingItem,
> ) (exec.Node, error) {
> ```

Composing the planning like this might open new doors for testing. I don't think much testing was ever added for distsql physical planning. Can you think of some tests that might be worth adding?
___
*[pkg/sql/distsql_spec_exec_factory.go, line 74 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/49682#-M8yePy84VE9sLl2ZOY6:-M8yePy84VE9sLl2ZOY7:b-qfyeq2) ([raw file](https://github.com/cockroachdb/cockroach/blob/429f61ae29994738abe4d8fd2f09c5c2f45b30ee/pkg/sql/distsql_spec_exec_factory.go#L74)):*
> ```Go
> 	recommendation := canDistribute
> 
> 	// ------- begin execFactory.ConstructScan -------
> ```

Are these comments convention in `execFactory`? `s/execFactory/distSQLSpecExecFactory`
___
*[pkg/sql/distsql_spec_exec_factory.go, line 100 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/49682#-M8yf6wd0HEFqz3UOVVi:-M8yf6wd0HEFqz3UOVVj:b-2ows2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/429f61ae29994738abe4d8fd2f09c5c2f45b30ee/pkg/sql/distsql_spec_exec_factory.go#L100)):*
> ```Go
> 	// TODO(yuzefovich): scanNode adds "parallel" attribute in walk.go when
> 	// scanNode.canParallelize() returns true. We should plumb that info from
> 	// here somehow as well.
> ```

What information is missing in this scope to do that? I think you have `maxResults`, limit information and you would just need to check the cluster setting
___
*[pkg/sql/exec_factory_util.go, line 34 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/49682#-M8ydnwJ5kw-U0W4vPKP:-M8ydnwJ5kw-U0W4vPKQ:b-dewv1q) ([raw file](https://github.com/cockroachdb/cockroach/blob/aecdd6291256c766517e8ee1960788634ff495e1/pkg/sql/exec_factory_util.go#L34)):*
> ```Go
> 		instrumentation: planner.curPlan.instrumentation,
> 	}
> 	assignPlan := func(plan *planMaybePhysical, node exec.Node) {
> ```

Instead of passing a pointer to `planMaybePhysical`, it might be nicer to just take a `*planTop` here, pass in `res`, and have this function set the `main` field
___
*[pkg/sql/exec_factory_util.go, line 36 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/49682#-M8ydScL7SEKMCXJPaJY:-M8ydScL7SEKMCXJPaJZ:bv0isps) ([raw file](https://github.com/cockroachdb/cockroach/blob/aecdd6291256c766517e8ee1960788634ff495e1/pkg/sql/exec_factory_util.go#L36)):*
> ```Go
> 	assignPlan := func(plan *planMaybePhysical, node exec.Node) {
> 		if usesPlanNodeRepresentation {
> 			plan.planNode = node.(planNode)
> ```

If we're doing a type assertion, I think we can probably remove the boolean parameter and test for whatever type we want.
___
*[pkg/sql/plan_opt.go, line 182 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/49682#-M8yi50P-lIFlVsGDQid:-M8yi50P-lIFlVsGDQie:b-m3gfya) ([raw file](https://github.com/cockroachdb/cockroach/blob/429f61ae29994738abe4d8fd2f09c5c2f45b30ee/pkg/sql/plan_opt.go#L182)):*
> ```Go
> 		bld  *execbuilder.Builder
> 	)
> 	dsp := p.DistSQLPlanner()
> ```

When would `dsp` be nil?
___
*[pkg/sql/plan_opt.go, line 184 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/49682#-M8yiCWW1seWK4AITeDr:-M8yiCWW1seWK4AITeDs:br761ij) ([raw file](https://github.com/cockroachdb/cockroach/blob/429f61ae29994738abe4d8fd2f09c5c2f45b30ee/pkg/sql/plan_opt.go#L184)):*
> ```Go
> 	dsp := p.DistSQLPlanner()
> 	if mode := p.SessionData().ExperimentalDistSQLPlanningMode; dsp != nil && mode != sessiondata.ExperimentalDistSQLPlanningOff {
> 		bld = execbuilder.New(newDistSQLSpecExecFactory(p, dsp), execMemo, &opc.catalog, root, p.EvalContext())
> ```

One thing that I'm wondering is why we create a new exec factory for every query. Is that expensive? Maybe we'd want a sync.pool, but not in this PR.

**Review by yuzefovich — COMMENTED** [review]
Hm, differential testing requires that both approaches produce exactly the same physical plans, but I think this will not be the case because we're changing the "distribution" strategy - in a follow-up PR I want to introduce the notion of "partially distributed" plans. And the old path would produce "local" plans while the new path will be producing "partially distributed" plans, so the differential testing is not applicable in general.

I think the best testing strategy will be to invest into a comprehensive suite of logic tests that use `EXPLAIN (distsql)`. A nice addition to this approach would be introducing a textual representation of the diagram so that it would be easier to examine (similar to what we have with `EXPLAIN (vec)`). I think we might have an issue for that.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49682#-:-M9-a9_p6v2V16eZkLTj:bvwe2q9)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto and @rytaft)

___
*[pkg/sql/distsql_physical_planner.go, line 911 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49682#-M8uOxo43lvgQ2PaVCwr:-M8zv6qD17i5izK2vSWo:bt1b19p) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc5e82a0abff1d13aca14648df0cc9fac7207307/pkg/sql/distsql_physical_planner.go#L911)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: can this function name be clearer? maybe just `getColOrdinals`?
</blockquote></details>

Renamed.
___
*[pkg/sql/distsql_physical_planner.go, line 1063 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/49682#-M8yeCct1fsKD9nY84L9:-M8zu09a2e5oO0MbK5Ff:b1qa1sr) ([raw file](https://github.com/cockroachdb/cockroach/blob/429f61ae29994738abe4d8fd2f09c5c2f45b30ee/pkg/sql/distsql_physical_planner.go#L1063)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

These are a lot of arguments. What do you think of creating some argument struct and maybe documenting the arguments better?
</blockquote></details>

I extracted a utility struct, but I don't think adding some documentation is that useful because most of the fields are simply extracted from `scanNode`s, so I referred the reader to the documentation in there. Or is your comment about improving that documentation in general as well?
___
*[pkg/sql/distsql_spec_exec_factory.go, line 60 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/49682#-M8yhDhY9Iramwdn9Mct:-M9-WWBxEdOPi14YzSiB:b-2s5qwz) ([raw file](https://github.com/cockroachdb/cockroach/blob/429f61ae29994738abe4d8fd2f09c5c2f45b30ee/pkg/sql/distsql_spec_exec_factory.go#L60)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Composing the planning like this might open new doors for testing. I don't think much testing was ever added for distsql physical planning. Can you think of some tests that might be worth adding?
</blockquote></details>

Hm, having some kind of unit tests doesn't seem that much useful, but using `EXPLAIN (distsql)` more to test the physical planning is probably the way to go. I think we could expand the "execbuilder" logic tests (i.e. `testoptlogic`) to include the physical planning once the new factory is significantly implemented.
___
*[pkg/sql/distsql_spec_exec_factory.go, line 74 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/49682#-M8yePy84VE9sLl2ZOY6:-M9-XMKt6jHw1Q8F-Za5:b48e5e6) ([raw file](https://github.com/cockroachdb/cockroach/blob/429f61ae29994738abe4d8fd2f09c5c2f45b30ee/pkg/sql/distsql_spec_exec_factory.go#L74)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Are these comments convention in `execFactory`? `s/execFactory/distSQLSpecExecFactory`
</blockquote></details>

I left them mostly for the ease of reviewing (and for myself when writing the code). I don't think they provide much benefit when the PR is ready to be merged, so I removed them.
___
*[pkg/sql/distsql_spec_exec_factory.go, line 100 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/49682#-M8yf6wd0HEFqz3UOVVi:-M9-Xl8wC7gY9gkyBRO2:bsvdg5i) ([raw file](https://github.com/cockroachdb/cockroach/blob/429f61ae29994738abe4d8fd2f09c5c2f45b30ee/pkg/sql/distsql_spec_exec_factory.go#L100)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

What information is missing in this scope to do that? I think you have `maxResults`, limit information and you would just need to check the cluster setting
</blockquote></details>

We do have all the necessary information in this scope, the complication is plumbing that information into `EXPLAIN` code which is currently not supported on the new representation (because it assumes walking over `planNode` tree).
___
*[pkg/sql/exec_factory_util.go, line 34 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/49682#-M8ydnwJ5kw-U0W4vPKP:-M9-ZHrA5oQbj5nTlZI-:b4h4ry6) ([raw file](https://github.com/cockroachdb/cockroach/blob/aecdd6291256c766517e8ee1960788634ff495e1/pkg/sql/exec_factory_util.go#L34)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Instead of passing a pointer to `planMaybePhysical`, it might be nicer to just take a `*planTop` here, pass in `res`, and have this function set the `main` field
</blockquote></details>

This function is also used for subqueries and cascades which are `planMaybePhysical` and not `planTop`, so I don't think we can do that.
___
*[pkg/sql/exec_factory_util.go, line 36 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/49682#-M8ydScL7SEKMCXJPaJY:-M9-a79Q5a0SBMgJl7XA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aecdd6291256c766517e8ee1960788634ff495e1/pkg/sql/exec_factory_util.go#L36)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

If we're doing a type assertion, I think we can probably remove the boolean parameter and test for whatever type we want.
</blockquote></details>

Done.
___
*[pkg/sql/plan_opt.go, line 182 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/49682#-M8yi50P-lIFlVsGDQid:-M9-YIIa63zEtFZhc64g:br59p23) ([raw file](https://github.com/cockroachdb/cockroach/blob/429f61ae29994738abe4d8fd2f09c5c2f45b30ee/pkg/sql/plan_opt.go#L182)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

When would `dsp` be nil?
</blockquote></details>

Indeed, it seems like it is always set, so it was an unnecessary "defense" mechanism, removed.
___
*[pkg/sql/plan_opt.go, line 184 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/49682#-M8yiCWW1seWK4AITeDr:-M9-YXiS9e-u-lkbrZ1d:b-mcenjn) ([raw file](https://github.com/cockroachdb/cockroach/blob/429f61ae29994738abe4d8fd2f09c5c2f45b30ee/pkg/sql/plan_opt.go#L184)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

One thing that I'm wondering is why we create a new exec factory for every query. Is that expensive? Maybe we'd want a sync.pool, but not in this PR.
</blockquote></details>

Hm, interesting point, I don't see a reason why we couldn't use a pool for this, but I'm also not sure whether it'll provide noticeable performance change. Left a TODO to look into this.

**Review by asubiotto — APPROVED** [review]
In that case, could we run differential testing without the distribution variable (i.e. always plan locally)? I think we need to come up with some way to make sure that we're not introducing bugs with this new planning approach.

I'm not sure I understand the `EXPLAIN (distsql)` approach, wouldn't the distribution still be an issue? Or do you mean that we wouldn't compare it with anything and write new tests for this new representation? That could work, and might be good to have, but it seems like a pretty heavy investment and doesn't mean we can't do some type of differential testing.

:lgtm: 

> Reviewed 10 of 10 files at r8, 5 of 5 files at r9, 6 of 6 files at r10, 5 of 5 files at r11.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49682#-:-M9J22IY5eGV2DcU4AP2:b-fn1fbj)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @rytaft and @yuzefovich)

___
*[pkg/sql/distsql_physical_planner.go, line 1063 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/49682#-M8yeCct1fsKD9nY84L9:-M9J3GMx4ESEd0nxTZqd:bgbin6g) ([raw file](https://github.com/cockroachdb/cockroach/blob/429f61ae29994738abe4d8fd2f09c5c2f45b30ee/pkg/sql/distsql_physical_planner.go#L1063)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I extracted a utility struct, but I don't think adding some documentation is that useful because most of the fields are simply extracted from `scanNode`s, so I referred the reader to the documentation in there. Or is your comment about improving that documentation in general as well?
</blockquote></details>

My comment was also about improving documentation in general as well but this LGTM.
___
*[pkg/sql/distsql_spec_exec_factory.go, line 74 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/49682#-M8yePy84VE9sLl2ZOY6:-M9J3w7_-9pd-D_vdHiz:bfqie4m) ([raw file](https://github.com/cockroachdb/cockroach/blob/429f61ae29994738abe4d8fd2f09c5c2f45b30ee/pkg/sql/distsql_spec_exec_factory.go#L74)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I left them mostly for the ease of reviewing (and for myself when writing the code). I don't think they provide much benefit when the PR is ready to be merged, so I removed them.
</blockquote></details>

I actually kind of liked the general comments describing the stages minus the format
___
*[pkg/sql/distsql_spec_exec_factory.go, line 100 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/49682#-M8yf6wd0HEFqz3UOVVi:-M9J4D-51pP09O_ZkIF3:b-wurtk1) ([raw file](https://github.com/cockroachdb/cockroach/blob/429f61ae29994738abe4d8fd2f09c5c2f45b30ee/pkg/sql/distsql_spec_exec_factory.go#L100)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

We do have all the necessary information in this scope, the complication is plumbing that information into `EXPLAIN` code which is currently not supported on the new representation (because it assumes walking over `planNode` tree).
</blockquote></details>

Ok. Just to confirm,  the tableReader still parallelizes scans when it can with the current code?

**Review by yuzefovich — COMMENTED** [review]
> In that case, could we run differential testing without the distribution variable (i.e. always plan locally)? I think we need to come up with some way to make sure that we're not introducing bugs with this new planning approach.

Yes, we could. To be clear, my comment about inability to differential testing was "in general" (I was thinking about creating some randomized queries and comparing the plans on them), but I do intend to use the differential testing approach when applicable. Namely, I'm thinking that we might want to introduce a separate cluster setting for testing purposes that would prohibit partially distributed plans - whenever the new factory creates a partially distributed plan, if that setting is set, then we would fallback to the old path, but local and fully-distributed plans should be exactly the same in the old and the new paths. I also think that we will introduce non-default logic test mode to be used for different already existing "explain" logic tests.

> Or do you mean that we wouldn't compare it with anything and write new tests for this new representation?

Yes, I think that we will need to write separate tests for partially distributed plans because we haven't had such precedent before, so there is nothing to reuse. However, we will be able to reuse already existing tests if they don't end up with a partially distributed plan.

TL;DR is that I am actively thinking about the testing, and the follow-up PR implements some EXPLAIN variants in the new factory which makes testing possible.

TFTRs!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49682#-:-M9JVsfODgIqN9soa6VE:b-eewo50)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @asubiotto and @rytaft)

___
*[pkg/sql/distsql_physical_planner.go, line 1063 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/49682#-M8yeCct1fsKD9nY84L9:-M9JVcCQDM1mBxa_dtdX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/429f61ae29994738abe4d8fd2f09c5c2f45b30ee/pkg/sql/distsql_physical_planner.go#L1063)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

My comment was also about improving documentation in general as well but this LGTM.
</blockquote></details>

Done.
___
*[pkg/sql/distsql_spec_exec_factory.go, line 74 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/49682#-M8yePy84VE9sLl2ZOY6:-M9JUN9q3hHIJgHZGWBo:b-hfw54i) ([raw file](https://github.com/cockroachdb/cockroach/blob/429f61ae29994738abe4d8fd2f09c5c2f45b30ee/pkg/sql/distsql_spec_exec_factory.go#L74)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I actually kind of liked the general comments describing the stages minus the format
</blockquote></details>

Ok, added the modified comments back.
___
*[pkg/sql/distsql_spec_exec_factory.go, line 100 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/49682#-M8yf6wd0HEFqz3UOVVi:-M9JVJgh4kGxc9OblXRG:b-f7rppz) ([raw file](https://github.com/cockroachdb/cockroach/blob/429f61ae29994738abe4d8fd2f09c5c2f45b30ee/pkg/sql/distsql_spec_exec_factory.go#L100)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Ok. Just to confirm,  the tableReader still parallelizes scans when it can with the current code?
</blockquote></details>

Yes, the "parallelization" logic is done by `DistSQLPlanner` in `planTableReaders` which is shared with the old path, and this comment refers to only not being able to show the "parallel" attribute in the output of EXPLAIN.
